### PR TITLE
[Client-gen] Add services/status path in API server, and let the service controller use UpdateStatus() method

### LINKED ISSUE #19504: Shall we create subresource path for ServiceStatus
For top-level API types that have a `Status` field, we usually create a subresource path for it on the API server, and make an `UpdateStatus()` method in the client library. The `Service.Status` is the only exception, it doesn't have either of them. Is there a particular reason to break this convention?

I raised the issue because in #19349 we are autogenerating the UpdateStatus() method for the client library. The generator generates the `UpdateStatus()` method for the [client for Service](https://github.com/kubernetes/kubernetes/pull/19349/files#diff-6e7ec0db5816c399b12106abaad01f8bR80), but because there is no subresource path registered at the apiserver, that function will always fail.

@thockin @kubernetes/goog-cluster @nikhiljindal

### PR DESCRIPTION
**Pull request by caesarxuchao**
Fixes #19504 
@nikhiljindal @bprashanth

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/19541)

### CODE REVIEW

File: pkg/client/unversioned/services.go [hunk 1]
```diff
@@ -82,6 +83,13 @@ func (c *services) Update(svc *api.Service) (result *api.Service, err error) {
 	return
 }
 
+// UpdateStatus takes the name of the service and the new status.  Returns the server's representation of the service, and an error, if it occurs.
```
Comment (bprashanth) [inline pkg/client/unversioned/services.go]: UpdateStatus takes a Service object with the new status and applies it as an update to the existing Service.

File: pkg/registry/service/strategy.go [hunk 2]
```diff
@@ -118,3 +118,22 @@ func MatchServices(label labels.Selector, field fields.Selector) generic.Matcher
...
 }
+
+type serviceStatusStrategy struct {
+	svcStrategy
+}
+
+var StatusStrategy = serviceStatusStrategy{Strategy}
+
+// PrepareForUpdate clears fields that are not allowed to be set by end users on update of status
+func (serviceStatusStrategy) PrepareForUpdate(obj, old runtime.Object) {
```
Comment (bprashanth) [inline pkg/registry/service/strategy.go]: please add a TestServiceStatusStrategy in strategy_test that verifies that spec is overwritten

File: pkg/registry/service/strategy.go [hunk 3]
```diff
@@ -118,3 +118,22 @@ func MatchServices(label labels.Selector, field fields.Selector) generic.Matcher
 func ServiceToSelectableFields(service *api.Service) fields.Set {
 	return generic.ObjectMetaFieldsSet(service.ObjectMeta, true)
 }
+
+type serviceStatusStrategy struct {
+	svcStrategy
+}
+
+var StatusStrategy = serviceStatusStrategy{Strategy}
```
Comment (bprashanth) [inline pkg/registry/service/strategy.go]: please add a godoc for exported variable:
// StatusStrategy is the default logic invoked when creating and updating Services.

**Comment by bprashanth** [comment]
Thanks, just nits

**Comment by caesarxuchao** [comment]
Comments addressed. PTAL. Thanks.

**Comment by bprashanth** [comment]
LGTM please squash and apply the label yourself once that's done

**Comment by caesarxuchao** [comment]
Add the label per previous comment.

**Comment by caesarxuchao** [comment]
@k8s-bot e2e test this please

File: test/e2e/service.go [hunk 4]
```diff
@@ -576,7 +576,9 @@ var _ = Describe("Services", func() {
 		})
 		Expect(err).NotTo(HaveOccurred())
 
-		if len(service.Status.LoadBalancer.Ingress) != 0 {
+		// Updating the service type shouldn't change the Status immediately. The status should be
+		// updated after waitForLoadBalancerDestroy.
+		if len(service.Status.LoadBalancer.Ingress) == 0 {
```
Comment (caesarxuchao) [inline test/e2e/service.go]: @bprashanth could you take a look at this e2e change?

Comment (caesarxuchao) [inline test/e2e/service.go]: @bprashanth a friendly reminder.

**Comment by bprashanth** [comment]
LGTM. @kubernetes/goog-cluster I wonder if anything else will break because it relies on updating the stauts silently through /services endpoint. Only one way to find out.

**Comment by caesarxuchao** [comment]
unit test flake: [url](
https://00e9e64bac0daea2026473b55fe30296a36280734685788ae4-apidata.googleusercontent.com/download/storage/v1_internal/b/kubernetes-jenkins/o/pr-logs%2Fb6293a68ea810797cc01fb7a38a1cd9bf28b48fc%2Fkubernetes-pull-test-unit-integration%2F10872%2Fbuild-log.txt?qk=AD5uMEsNq6-GyyD9JTmUFhLSpNAM3tEO-Sy5VBk_b3Pn1KF8FkVA_SrrKi3BjhHKevQe1q0bkv8AiFLfyxzCi9GX322w0JAJY_3f9rCpsymReWBhe-oLuuSiEwba8RLQpOeZX7DCr_8o9mId_PpL-pk9ltLheMAaGW9MF6ZhMEZ6P_ysPBZU-QUk_t0ODKV04P_KzfNOTjt5hefhVIHFOH0bdgKcYOhMkWkPWDCXqJqbQWkcMoBvL6D4pz_hDi-hv8swV62KgYWNAplppgdRJ3JinVycwzXBdjdN84H1UCjezCDUJdeCyzuX7nE2zY5bEr7fG1zj9EEaqvieFkY4vIOjpanLmYQBJdTz0_J8ZugKdRHPr2JMfO2aCE_jADJ7ufaBZzR1W_5HfBWjagGW3US4WznIn4fLRf7AF5M0T27eReVKs3xamKSEfgVMEBNwNkqX-ue9GDrwZpjJ7bcjumi8rJ5fGiNfK2Y6Z9IjAkORm1LOhqlS6Nk4vMgFTQtxARjmAEL_6Uq5wZ7nYq81Tl0Wf8hergofya7-aiBu2u1ipzOP3JOJZy0EoClPzR7AQTSJDFyHZWXk-0ajM9SJEKKMD3b8AiUbh_09YeQESfJVuwVPFNGYhNA1rKY02wqN0xTgDmF0DwShpQ0OXbMO4NC69zquLbOFeNv_Wbx3-2zC8Tv5m_ENMY91a8tXyO0eGsYzca0nday_wwQbUVm8grxNxkfbgCdIRpGEVd8-Tqq5vkJlkxhkw5yUlGgb2OY2NrjdG4p33Y3CbFoE9uIVjDTkujglerU0lEH-7slsdLreD4L4hs-lxuMxTx7e20hXjX-lPPf0ouoL58puasheijWC5Ob-g1d6Re5wbGxjzBAx1v6QW6ALlrLmBQIctSrgSpLbB8gA0z4NHsqG-1JHn4-kFibmZNeZnYF0hubhCWSbzoKpkL4GILBzhBsq8swDJ8Dawxee2LqhHfCzYdg6mksyKqQwSJtRng)

**Comment by caesarxuchao** [comment]
e2e also seems to be flake.

**Comment by caesarxuchao** [comment]
@k8s-bot test this please

**Comment by caesarxuchao** [comment]
e2e might be a flake:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/networking.go:50
Jan 31 13:58:33.077: Unable to connect/talk to the internet: Get http://www.google.com/: read tcp 10.240.0.7:45476->209.85.146.105:80: read: connection reset by peer
```
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/19541/kubernetes-pull-build-test-e2e-gce/25930/?debugUI=DEVELOPERS

**Comment by caesarxuchao** [comment]
@k8s-bot e2e test this please

**Pull request merged**